jsonschema/tests/test_validators.py

```diff
@@ -1194,6 +1194,21 @@ def test_validator_for_jsonschema_default(self):
     def test_validator_for_custom_default(self):
         self.assertIs(validators.validator_for({}, default=None), None)
 
+    def test_validator_warns_if_schema_not_found(self):
+        self.assertWarns(
+            category=DeprecationWarning,
+            message=(
+                "This schema was not found but going to validate with latest draft. "
+                "This will raise a SchemaError in future. "
```

Comment:
What schema are you referring to here - in `this schema` . `"This schema defines a $schema property for an unregistered or unknown meta-schema.`

Reply:
The one that was passed into either `validator_for` or transitively `validates`.

Reply:
fixed the other comments. Good point about leaving it unspecified. And about the schema - you mean whatever is the value of this `schema.get(u"$schema", u"")` . I believe I get it now, but shouldn't we add the name of the schema we are saying is unregistered to the warning message?

Reply:
Literally in that line `schema` :)

The thing that *contains* the `$schema`.

This situation we're deprecating basically happens when someone says

`validator_for({"$schema", "foobarbaz", ...})` or `validate(someinstance, {"$schema", "foobarbaz", ...})`, where foobarbaz is supposedly the metaschema for that schema, but didn't tell `jsonschema` which validator should be used to validate schemas that use that.

Which actually I believe points out a bug in the implementation here.

You're calling `schema.get("$schema", "")` and warning any time that the resulting `$schema` isn't registered, but that will warn even in the case where `$schema` isn't present. Doing that isn't wise, but isn't per se part of what we need to deprecate here. The thing we're deprecating here is the behavior that takes a schema like `{"$schema": "unknown"}` and *silently* still uses a validator for that, rather than saying "I don't know what meta schema `unknown` refers to, so I can't pick a Validator class for it".

Does that make sense?